libs/elasticsearch-nio/src/main/java/org/elasticsearch/nio/BytesChannelContext.java

```diff
@@ -77,19 +119,35 @@ public boolean hasQueuedWriteOps() {
     }
 
     @Override
-    public void clearQueuedWriteOps(Exception e) {
-        assert channel.getSelector().isOnCurrentThread() : "Must be on selector thread to clear queued writes";
-        for (WriteOperation op : queued) {
-            channel.getSelector().executeFailedListener(op.getListener(), e);
+    public void closeChannel() {
+        if (isClosing.compareAndSet(false, true)) {
+            channel.getSelector().queueChannelClose(channel);
+        }
+    }
+
+    @Override
+    public boolean readyToClose() {
+        return peerClosed || ioException || isClosing.get();
+    }
+
+    @Override
+    public void closeFromSelector() {
+        assert channel.getSelector().isOnCurrentThread() : "Must be on selector thread to close";
+        isClosing.set(true);
+        channelBuffer.close();
+        for (BytesWriteOperation op : queued) {
+            channel.getSelector().executeFailedListener(op.getListener(), new ClosedChannelException());
         }
         queued.clear();
     }
 
-    private void singleFlush(WriteOperation headOp) throws IOException {
+    private void singleFlush(BytesWriteOperation headOp) throws IOException {
         try {
-            headOp.flush();
+            int written = channel.write(headOp.getBuffersToWrite());
+            headOp.incrementIndex(written);
         } catch (IOException e) {
             channel.getSelector().executeFailedListener(headOp.getListener(), e);
+            ioException = true;
```

Comment:
does `ioException` need to be volatile?

Reply:
Not. Read, write, and checking if the channel is ready for close are only called from the selector thread. The only channel context methods that can be called from a different thread are `sendMessage` and `closeChannel`. I eventually intend to follow up a change the cleans this relationship up. I'll add a comment explaining why it is not volatile.